File: Java/Algorithm/Algorithm I/120. Triangle.java
import java.util.List;

class Solution {

  /**
   * Calculates the minimum path sum from the top to the bottom of a triangle.
   *
   * @param triangle A List of Lists representing the triangle with integer values.
   * @return The minimum path sum.
   */
  public int minimumTotal(List<List<Integer>> triangle) {
    int ROWS = triangle.size();

    for (int row = ROWS - 2; row >= 0; row--) {
      int COLUMNS = triangle.get(row).size();

      for (int column = 0; column < COLUMNS; column++) {
        int currentCellValue = triangle.get(row).get(column);
        int nextRowValue1 = triangle.get(row + 1).get(column);
        int nextRowValue2 = triangle.get(row + 1).get(column + 1);

        triangle
          .get(row)
          .set(
            column,
            currentCellValue + Math.min(nextRowValue1, nextRowValue2)
          );
      }
    }

    return triangle.get(0).get(0);
  }
}

File: Python/Algorithm/Algorithm I/120. Triangle.py
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        """
        Find the minimum path sum from top to bottom in a triangle.

        Args:
            triangle (List[List[int]]): A list of lists representing the triangle.

        Returns:
            int: The minimum path sum.

        Time Complexity: o(n) where n is the number of elements in the triangle.
        Space Complexity: o(1)
        """
        ROWS = len(triangle)

        for row in range(ROWS - 2, -1, -1):
            COLUMNS = len(triangle[row])

            for column in range(COLUMNS):
                firstOption = triangle[row + 1][column]
                secondOption = triangle[row + 1][column + 1]

                triangle[row][column] += min(firstOption, secondOption)

        return triangle[0][0]
